crates/ra_hir_ty/src/traits.rs

```diff
@@ -39,16 +41,36 @@ impl TraitSolver {
         goal: &chalk_ir::UCanonical<chalk_ir::InEnvironment<chalk_ir::Goal<TypeFamily>>>,
     ) -> Option<chalk_solve::Solution<TypeFamily>> {
         let context = ChalkContext { db, krate: self.krate };
-        debug!("solve goal: {:?}", goal);
+        log::debug!("solve goal: {:?}", goal);
         let mut solver = match self.inner.lock() {
             Ok(it) => it,
             // Our cancellation works via unwinding, but, as chalk is not
             // panic-safe, we need to make sure to propagate the cancellation.
             // Ideally, we should also make chalk panic-safe.
             Err(_) => ra_db::Canceled::throw(),
         };
-        let solution = solver.solve(&context, goal);
-        debug!("solve({:?}) => {:?}", goal, solution);
+
+        let solution = panic::catch_unwind({
+            let solver = panic::AssertUnwindSafe(&mut solver);
+            let context = panic::AssertUnwindSafe(&context);
+            move || solver.0.solve(context.0, goal)
+        });
+
+        let solution = match solution {
+            Ok(it) => it,
+            Err(err) => {
+                if err.downcast_ref::<Canceled>().is_some() {
+                    panic::resume_unwind(err)
+                } else {
+                    log::error!("chalk panicked :-(");
```

Comment:
does this mean we won't get a stacktrace at all? :confused:

Reply:
Nope. After a panic, the first thing that happens is that the panic hook is invoked. The panic hook prints the backtrace. After that, the stack-unwinding process begins.

In other words, panics don't carry backtraces with them, rather, backtrace is printed immediately, while all the frames are still on the stack.

Reply:
Ah, ok.